### DISCUSSION TITLE
Using opt-in model for NuGet updates (non-major)

### DISCUSSION THREAD

**Discussion by tomkerkhove** [renovatebot/renovate#17934] [Request Help]
Hi,

I'm trying to setup a configuration where all .NET/NuGet packages are ignored by default unless explicitly opted-in. As part of this, we only want to have the non-major updates.

So I came up with the following configuration:
```json
{
  "$schema": "https://docs.renovatebot.com/renovate-schema.json",
  "separateMajorMinor": true,
  "labels": ["dependencies"],
  "ignoreDeps": [],
  "printConfig": true,
  "semanticCommits": "enabled",
  "prHourlyLimit": 0,
  "prConcurrentLimit": 0,
  "enabledManagers": ["dockerfile", "docker-compose", "nuget"],
  "packageRules": [
    {
      "datasources": ["nuget"],
      "groupName": "Ignore NuGet updates out-of-the-box",
      "enabled": false
    },
    {
      "datasources": ["docker"],
      "groupName": "Docker images"
    },
    {
      "matchPackagePrefixes": ["mcr.microsoft.com/dotnet/"],
      "groupName": ".NET Core Docker containers"
    },
    {
      "matchPackagePrefixes": ["Internal"],
      "matchUpdateTypes": ["minor", "patch"],
      "groupName": "Internal Framework",
      "enabled": true
    },
    {
      "datasources": ["nuget"],
      "matchPackagePrefixes": ["AutoMapper"],
      "groupName": "AutoMapper",
      "enabled": true
    }
  ],
  "extends": [
    "config:base",
    "docker:pinDigests"
  ],
  "hostRules": [
    {
      "hostType": "nuget",
      "matchHost": "pkgs.dev.azure.com",
      "username": "",
      "password": "<redacted>",
      "abortOnError": true
    }
  ]
}
```

However, the onboarding PR is only detecting the AutoMapper updates and skips the internal framework with `"skipReason": "disabled"`.

Is there another way to achieve this or am I doing something wrong?

**Comment by rarkins** [comment]
Your first rule stops them from even getting looked up.

Do something like this:

```
    {
      "description": "Disable all nuget packages except Internal",
      "matchDatasources": ["nuget"],
      "excludePackagePrefixes": ["Internal"],
      "enabled": false
    },
    {
      "description": "Disable major nuget updates",
      "matchDatasources": ["nuget"],
      "matchUpdateTypes": ["major"],
      "enabled": false
    },
    // Now apply your grouping rule to what's left
```

**Reply by tomkerkhove** [reply]
> Your first rule stops them from even getting looked up.

Not exactly because `AutoMapper` is getting maintained, while the other one is not. But let me try your suggestion!

**Reply by rarkins** [reply]
You're right, and that's expected - I just didn't look that far. I'll explain in more detail: you have a rule to disable all nuget packages, and then later one to re-enable AutoMapper. That works. You also have a rule to re-enable minor/patch updates of Internal packages. But that doesn't work, because Internal packages were all disabled before they could be looked up. So if they weren't looked up, we couldn't know if there's patch or minor available, and therefore cannot re-enable them.

**Reply by tomkerkhove** [reply]
No luck:
![image](https://user-images.githubusercontent.com/4345663/191989082-6d466b8a-f55b-41ce-ade6-57a37a207e6c.png)

Config:
```json
"packageRules": [
    {
      "groupName": "Ignore NuGet updates out-of-the-box",
      "matchDatasources": ["nuget"],
      "excludePackagePrefixes": ["AutoMapper", "Internal"],
      "enabled": false
    },
    {
      "description": "Disable major NuGet updates",
      "matchDatasources": ["nuget"],
      "matchUpdateTypes": ["major"],
      "enabled": false
    },
    {
      "datasources": ["docker"],
      "groupName": "Docker images"
    },
    {
      "matchPackagePrefixes": ["mcr.microsoft.com/dotnet/"],
      "groupName": ".NET Core Docker containers"
    },
    {
      "matchPackagePrefixes": ["Internal"],
      "groupName": "Internal Framework",
      "enabled": true
    },
    {
      "datasources": ["nuget"],
      "matchPackagePrefixes": ["AutoMapper"],
      "groupName": "AutoMapper",
      "enabled": true
    }
  ]
```

I'll try again by explicitly adding `"matchUpdateTypes": ["minor", "patch"],` again.

**Reply by tomkerkhove** [reply]
> I'll try again by explicitly adding `"matchUpdateTypes": ["minor", "patch"],` again.

Ok, this did it - Thank you for the help!

Also, Renovate is superduperfast! Big kudos for building it; saves a ton of time
